src/vm/hooks.rs

```diff
@@ -0,0 +1,199 @@
+//! VM hooks
+//!
+//! Make it possible to execute custom arbitrary code at different stages of the VM execution
+//!
+//! If added to the VM, hooks function will be called during the VM execution at specific stages.
+//!
+//! Available hooks:
+//! - before_first_step, executed before entering the execution loop in [run_until_pc](CairoRunner::run_until_pc)
+//! - pre_step_instruction, executed before each instruction_step in [step](VirtualMachine::step)
+//! - post_step_instruction, executed after each instruction_step in [step](VirtualMachine::step)
+
+use std::{any::Any, collections::HashMap, sync::Arc};
+
+use felt::Felt;
+
+use crate::{
+    hint_processor::hint_processor_definition::HintProcessor, types::exec_scope::ExecutionScopes,
+};
+
+use super::{
+    errors::vm_errors::VirtualMachineError, runners::cairo_runner::CairoRunner,
+    vm_core::VirtualMachine,
+};
+
+type BeforeFirstStepHookFunc = Arc<
+    dyn Fn(
+            &mut VirtualMachine,
+            &mut CairoRunner,
+            &HashMap<usize, Vec<Box<dyn Any>>>,
+        ) -> Result<(), VirtualMachineError>
+        + Sync
+        + Send,
+>;
+
+type StepHookFunc = Arc<
+    dyn Fn(
+            &mut VirtualMachine,
+            &mut dyn HintProcessor,
+            &mut ExecutionScopes,
+            &HashMap<usize, Vec<Box<dyn Any>>>,
+            &HashMap<String, Felt>,
+        ) -> Result<(), VirtualMachineError>
+        + Sync
+        + Send,
+>;
+
+/// The hooks to be executed during the VM run
+///
+/// They can be individually ignored by setting them to [None]
+#[derive(Clone, Default)]
+pub struct Hooks {
+    before_first_step: Option<BeforeFirstStepHookFunc>,
+    pre_step_instruction: Option<StepHookFunc>,
+    post_step_instruction: Option<StepHookFunc>,
+}
+
+impl Hooks {
+    pub fn new(
+        before_first_step: Option<BeforeFirstStepHookFunc>,
+        pre_step_instruction: Option<StepHookFunc>,
+        post_step_instruction: Option<StepHookFunc>,
+    ) -> Self {
+        Hooks {
+            before_first_step,
+            pre_step_instruction,
+            post_step_instruction,
+        }
+    }
+}
+
+impl VirtualMachine {
+    pub fn execute_before_first_step(
+        &mut self,
+        runner: &mut CairoRunner,
+        hint_data_dictionary: &HashMap<usize, Vec<Box<dyn Any>>>,
+    ) -> Result<(), VirtualMachineError> {
+        if let Some(hook_func) = self.hooks.clone().before_first_step {
```

Comment:
We need to clone it because we are later using `self` (as an &mut) as an argument of the `hook_func` function. Meaning that, if not clone, the value of the `hook_func` could be changed in `self` while being executed.
The borrow checker really doesn't like this kind of stuff. So we first clone `hooks` in order to get a stable, unalterable, value for `before_first_step`.

```rust
 (hook_func)(
                self,
                hint_executor,
                exec_scope,
                hint_data_dictionary,
                constants,
            )?;
```